dali/driver/base.py

```diff
@@ -234,3 +236,66 @@ def close(self):
         self._disconnecting = True
         self._stop_listening.set()
         super(USBListener, self).close()
+
+class SerialBackend(Backend):
+    """``Backend`` implementation for Quad-DALI-USB-Interface. Uses serial connection."""
+
+    def __init__(self, port, baudrate=115200, bytesize=serial.EIGHTBITS, parity=serial.PARITY_NONE,
+                 stopbits=serial.STOPBITS_ONE):
+        """Open connection to the DALI interface.
+
+        @param port: valid serial port (e.g. /dev/ttyUSB0 or COM1)
+        @param baudrate: serial baudrate; default is 115.200
+        @param bytesize: number of bits per byte; default is 8
+        @param parity: configures parity bit; default is none
+        @param stopbits: number of stopbits per byte; default is one"""
+
+        self._serial = serial.Serial(port=port, baudrate=baudrate, bytesize=bytesize, parity=parity, stopbits=stopbits)
+
+        # for compatibility with older pyserial versions
+        # background: the methods for flushing the in-/out-buffer were renamed in v3.0
+        if serial.VERSION.split('.')[0] == '2':
+            self._reset_input_buffer = self._serial.flushInput
+            self._reset_output_buffer = self._serial.flushOutput
+        elif serial.VERSION.split('.')[0] == '3':
+            self._reset_input_buffer = self._serial.reset_input_buffer
+            self._reset_output_buffer = self._serial.reset_output_buffer
+        else:
+            raise RuntimeError(f'pyserial={serial.VERSION} is not supported')
+
+        # flush all buffers to remove old data before checking the port
+        self._reset_output_buffer()
+        sleep(0.050) # just a precaution if the previous command sent some data
+        self._reset_input_buffer()
+
+    def read(self, timeout=None):
+        """Read data from the DALI interface.
+
+        @param timeout: read timeout in seconds; set to 0 or None to disable
+        @return data read from the interface (bytes)"""
+
+        # internal read timeout is set to 10 ms to increase responsiveness
+        self._serial.timeout = 0.010
+        # to compensate for the short read timeout and to use the timeout argument, the read is repeated until
+        # more time than specified has elapsed
+        start = perf_counter()
+        data = bytes()
+        while timeout and abs(start-perf_counter()) < timeout:
```

Comment:
You can define the read timeout when creating the serial.Serial instance.

Reply:
That would be a deviation from ``Backend``.

I also think it is useful to be able to specify the ``timeout`` on a per-read basis. Of course, this could also be done using a public ``timeout`` attribute.

Reply:
I have not taken a look at the serial implementation. Maybe it's possible to override the timeout on the serial instance.

Reply:
https://github.com/pyserial/pyserial/blob/master/serial/serialutil.py#L359

it is possible to set the timeout on an instance. However, setting it causes _reconfigure_port to be called which looks like a heavy operation at a first glance.

 https://github.com/pyserial/pyserial/blob/master/serial/serialposix.py#L358

maybe a timeout can be held at the backend instance and set only if it changed.

Reply:
In my implementation the ``timeout`` of the serial connection is fixed and the actual timeout is handled by the backend code. I think this was done to make the code more responsive and alleviate issues with lost bytes.

I will move the assignment to the ``__init__`` method, so it's only done once.

Reply:
I added a note in #69. There is the question if we really need cycle based timeouts.